### DISCUSSION TITLE
Deploying to production - build is in local mode

### DISCUSSION THREAD

**Discussion by mAAdhaTTah** [tinacms/tinacms#3426] [Help] [answered]
I'm passing the client to defineConfig & am still getting the Nextjs build running in local-mode. PR I'm working on is here: https://github.com/mAAdhaTTah/frontend/pull/739 When running the build, I get this error:

```
FetchError: request to http://localhost:4001/graphql failed, reason: connect ECONNREFUSED 127.0.0.1:4001
--
12:22:26.033 | at ClientRequest.<anonymous> (/vercel/path0/node_modules/node-fetch/lib/index.js:1491:11)
12:22:26.033 | at ClientRequest.emit (node:events:513:28)
12:22:26.033 | at Socket.socketErrorListener (node:_http_client:494:9)
12:22:26.034 | at Socket.emit (node:events:513:28)
12:22:26.034 | at emitErrorNT (node:internal/streams/destroy:157:8)
12:22:26.034 | at emitErrorCloseNT (node:internal/streams/destroy:122:3)
12:22:26.034 | at processTicksAndRejections (node:internal/process/task_queues:83:21) {
12:22:26.034 | type: 'system',
12:22:26.034 | errno: 'ECONNREFUSED',
12:22:26.034 | code: 'ECONNREFUSED'
12:22:26.034 | }
12:22:26.034 |  
12:22:26.034 | > Build error occurred
12:22:26.036 | Error: Failed to collect page data for /gistpens/[slug]
12:22:26.036 | at /vercel/path0/node_modules/next/dist/build/utils.js:916:15
12:22:26.036 | at processTicksAndRejections (node:internal/process/task_queues:96:5) {
12:22:26.036 | type: 'Error'
12:22:26.037 | }
12:22:26.090 | Error: Command "npm run build" exited with 1
```

The Nextjs client should be using the production endpoints (which is def configured correctly, because the Tinacms build failed the first few times I ran it), right? How do I get this going?

**Comment by jamespohalloran** [comment]
Looks like you're getting a more sensible error in the CI build: https://github.com/mAAdhaTTah/frontend/actions/runs/3576782217/jobs/6015001242

> Error: Client not configured properly. Missing clientId, token.
 
If you're pulling the error you're describing from Netlify/Vercel, is there a chance your build in Netlify/Vercel is running a different script than `build`? 

> I'm passing the client to defineConfig
This should no longer be required

**Reply by mAAdhaTTah** [reply]
Yeah, that's a different problem because I haven't added the token to GitHub Actions. I'm actually going to remove that from CI because it isn't necessary.

The build is on Vercel. Def running `npm run build`:

![Screen Shot 2022-11-29 at 2 33 22 PM](https://user-images.githubusercontent.com/4371429/204630518-2e11a68d-6b85-4794-bde5-bb662e855602.png)

**Reply by mAAdhaTTah** [reply]
Also, in case I wasn't clear, this is failing while building Nextjs, using the generated Tina client, so on the `next build` step. `tinacms build` completes successfully.

**Comment by logan-anderson** [comment]
@mAAdhaTTah This does seem strange. I tried pulling down your repo and but the `npm install` fails. I tried a couple of different versions of node but could not get it to work. If you create a  small site (Maybe use the barebones starter `npx create-tina-app@latest` and the select barebones ) that reproduces this I would love to take a further look.

**Reply by mAAdhaTTah** [reply]
Yeah, I was getting "legacy peer deps" issues when installing, although I believe that should be resolved now. I do a lot of rebasing so you may need to delete & re-checkout the branch or rebase it yourself.

I'm running node v16.17.1 & npm 8.15.0 if that helps.

**Reply by logan-anderson** [reply]
I was able to get this working I see the issue and will post the solution in a separate thread

**Comment by logan-anderson** [comment, marked as answer]
This reason why you are getting this is because you are still using `ExperimentalGetTinaClient`. This assuming the local server is running and starting in [v.68.14](https://tina.io/blog/tina-v-0.68.14) it does not run in CI. To fix this you must use the new "Unified client" that is built by tinacms in `tinacms build`. 

Anywhere you do data [fetching tinacms you must use this client](https://tina.io/docs/data-fetching/overview/).

To fix this anywhere you have

```ts
const client = ExperimentalGetTinaClient();
const data = await client.<SomeQuery>();
```

Must be changed to

```ts
import { client } from '../[pathToTina]/.tina/__generated__/client'
// ...
const res = await client.queries.<SomeQuery>();
const data = res.data
const query = res.query
const variables = res.variables 
```

We should probably stop exporting `ExperimentalGetTinaClient` 

Please let me know if you have any questions.

**Reply by mAAdhaTTah** [reply]
Ah, that was it. Thanks!
